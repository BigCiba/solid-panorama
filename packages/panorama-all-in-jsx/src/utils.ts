import { NodePath } from '@babel/core';
import { addNamed } from '@babel/helper-module-imports';
import * as t from '@babel/types';

export function registerImportMethod(
    path: NodePath,
    name: string,
    moduleName: string
) {
    let imports = path.scope.getProgramParent().getData('imports');
    if (!imports) {
        imports = new Map();
        path.scope.getProgramParent().setData('imports', imports);
    }
    if (!imports.has(`${moduleName}:${name}`)) {
        let id = addNamed(path, name, moduleName, {
            nameHint: `_$${name}`
        });
        imports.set(`${moduleName}:${name}`, id);
        return id;
    } else {
        let iden = imports.get(`${moduleName}:${name}`);
        return t.cloneNode(iden);
    }
}
