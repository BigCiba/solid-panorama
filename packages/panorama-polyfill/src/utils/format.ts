// Based on https://github.com/browserify/node-util/blob/4b1c0c79790d9968eabecd2e9c786454713e200f/util.js#L33
export function format(value?: unknown, ...substitutions: unknown[]) {
    if (typeof value !== 'string') {
        return [value, ...substitutions].map(v => inspect(v)).join(' ');
    }

    let count = 0;
    let result = String(value).replace(/%[sdj%]/g, x => {
        count++;
        if (x === '%%') return '%';
        if (substitutions.length === 0) return x;

        switch (x) {
            case '%s':
                return String(substitutions.shift());

            case '%d':
                return String(Number(substitutions.shift()));

            case '%j':
                try {
                    return JSON.stringify(substitutions.shift());
                } catch {
                    return '[Circular]';
                }

            default:
                return x;
        }
    });

    for (const x of substitutions) {
        if (typeof x !== 'object' || x === null) {
            result += ` ${x}`;
        } else {
            result += ` ${inspect(x)}`;
        }
    }

    return result;
}

export function formatx(value?: unknown, ...substitutions: unknown[]) {
    if (typeof value !== 'string') {
        return [value, ...substitutions].map(v => inspect(v, true)).join(' ');
    }

    let result = String(value).replace(/%[sdj%]/g, x => {
        if (x === '%%') return '%';
        if (substitutions.length === 0) return x;

        switch (x) {
            case '%s':
                return String(substitutions.unshift());

            case '%d':
                return String(Number(substitutions.unshift()));

            case '%j':
                try {
                    return JSON.stringify(substitutions.unshift());
                } catch {
                    return '[Circular]';
                }

            default:
                return x;
        }
    });

    for (const x of substitutions) {
        if (typeof x !== 'object' || x === null) {
            result += ` ${x}`;
        } else {
            result += ` ${inspect(x)}`;
        }
    }

    return result;
}

function inspect(value: unknown, expand?: boolean, tab = '') {
    let result = '';
    if (typeof value === 'string') {
        result = `"${value}"`;
    } else if (typeof value === 'number' || typeof value === 'boolean') {
        result = `${value}`;
    } else if (typeof value === 'function') {
        result = isClass(value)
            ? `[class ${value.name}]`
            : `[function ${value.name}]`;
    } else if (typeof value === 'symbol') {
        result = value.toString();
    } else if (typeof value === 'undefined') {
        result = `undefined`;
    } else if (typeof value === 'bigint') {
        result = `[bigint ${value.toString()}]`;
    } else if (typeof value === 'object') {
        if (value === null) {
            result = 'null';
        } else if (Array.isArray(value)) {
            let list = [];
            for (const v of value) {
                list.push(
                    tab + inspect(v, expand, expand ? tab + '    ' : tab)
                );
            }
            if (expand) {
                result += '[\n';
                result += list.map(v => '    ' + v).join(',\n');
                result += '\n' + tab + ']';
            } else {
                result = `[ ${list.join(', ')} ]`;
            }
        } else {
            let list: string[] = [];
            let prefix = '';
            // Map
            if (value instanceof Map) {
                prefix = '[Map]';
                for (const [k, v] of value.entries()) {
                    let key = '';
                    if (typeof k === 'object') {
                        key = Array.isArray(k) ? '[Array]' : '[Object]';
                    } else if (k.toString) {
                        key = k.toString();
                    }
                    list.push(
                        `${tab}${key}: ${inspect(
                            v,
                            expand,
                            expand ? tab + '    ' : tab
                        )}`
                    );
                }
            } else if (value instanceof Set) {
                prefix = '[Set]';
                for (const v of value.values()) {
                    list.push(
                        `${tab}${inspect(
                            v,
                            expand,
                            expand ? tab + '    ' : tab
                        )}`
                    );
                }
            } else {
                const isPanel = isPanelBase(value);
                for (const [k, v] of Object.entries(value)) {
                    if (isPanel) {
                        console.log(value, k, v);
                    }
                    if (k === 'style' && isPanel) {
                        list.push(`${tab}${k}: [VCSSStyleDeclaration]`);
                        continue;
                    }
                    list.push(
                        `${tab}${k}: ${inspect(
                            v,
                            expand,
                            expand ? tab + '    ' : tab
                        )}`
                    );
                }
            }

            if (expand) {
                result += prefix + '{\n';
                result += list.map(v => '    ' + v).join(',\n');
                result += '\n' + tab + '}';
            } else {
                result = prefix + `{ ${list.join(', ')} }`;
            }
        }
    }
    return result;
}

const isPanelBase = (value: object): value is PanelBase =>
    'paneltype' in value &&
    'rememberchildfocus' in value &&
    'SetPanelEvent' in value;

function isClass(value: unknown) {
    if (typeof value !== 'function') return false;
    const descriptor = Object.getOwnPropertyDescriptor(value, 'prototype');
    if (!descriptor) return false;
    return !descriptor.writable;
}
