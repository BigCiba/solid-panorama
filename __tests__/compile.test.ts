import { describe, expect, test } from '@jest/globals';
import { parseSolid } from './utils';

describe('compile', function () {
    test('transform: normal', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';

            function Item(props) {
                const [rootStyle, setRootStyle] = createSignal('root');
                let visible = true
                let root, btnA
                return (
                    <Panel id="root" ref={root} className={rootStyle().join(' ')}>
                        <Label text="Testing" />
                        <Panel className="buttons">
                            <Button ref={(a) => btnA = a} />
                            <Button visible={visible} />
                            <Button />
                        </Panel>
                    </Panel>
                );
            }
            
            function HelloWorld() {
                return (
                    <Panel>
                        <Item />
                        <Item show />
                    </Panel>
                );
            }
            
            render(() => <HelloWorld />, $('#app'));        
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: fragment', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';
                    
            function HelloWorld() {
                return (
                    <>
                        <Button />
                        <Button />
                        <Button />
                    </>
                );
            }
            
            render(() => <HelloWorld />, $('#app'));  
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: for each', function () {
        const result = parseSolid(`
            import { render } from "solid-js/web";
            import { createSignal } from "solid-js";
            
            const App = () => {
                const [list, setList] = createSignal(['A']);
                
                function click() {
                    setList([...list(), "B"])
                }
                
                return <Panel>
                    <For each={list()} >
                        {(item, index) => {
                        return <Label text={item} />
                        }}
                    </For>
                    <Button onClick={click}>Click</Button>
                </Panel>;
            };
            
            render(() => <App />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: style', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';

            function Item(props) {
                return <Panel class={props.style} />
            }
            
            function HelloWorld() {
                return (
                    <Panel>
                        <Button style={{width:'12px', height: 12}} />
                        <Button style="width:12px; height: 12px" />
                        <Item style="red" />
                    </Panel>
                );
            }
            
            render(() => <HelloWorld />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: textNode', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';
            
            function HelloWorld() {
                return (
                    <Panel>
                        <Panel>
                            Welcome My Game
                        </Panel>
                        <Panel>
                            {\`<strong>Welcome</strong> My Game\`}
                        </Panel>
                        <Panel>
                            <Label text="Welcome" />
                            #addon_game_name
                            <Label text="(～￣▽￣)～" />
                        </Panel>
                    </Panel>
                );
            }
            
            render(() => <HelloWorld />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: snippet', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';
            
            function HelloWorld() {
                return (
                    <Button snippet="templateA" />
                );
            }
            
            render(() => <HelloWorld />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: spread operator', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';

            function Item(props) {
                return <Panel {...props} />
            }
            
            function HelloWorld() {
                return (
                    <Item />
                );
            }
            
            render(() => <HelloWorld />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: dynamic', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';

            const RedThing = () => <Panel style="color: red">Red Thing</Panel>;
            const GreenThing = () => <Panel style="color: green">Green Thing</Panel>;
            const BlueThing = () => <Panel style="color: blue">Blue Thing</Panel>;

            const options = {
                red: RedThing,
                green: GreenThing,
                blue: BlueThing
            }
            
            function HelloWorld() {
                const [selected, setSelected] = createSignal("red");
                return (
                    <Dynamic component={options[selected()]} />
                );
            }
            
            render(() => <HelloWorld />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('transform: vars', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';

            function HelloWorld() {
                const [ability, setAbility] = createSignal("");
                return (
                    <Panel>
                        <Panel vars={{name: 'robin'}} />
                        <Panel vars={{name: 'robin', ability: ability()}} dialogVariables={{}} />
                    </Panel>
                );
            }
            
            render(() => <HelloWorld />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('custom properties: initialize', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';

            function App() {
                let root

                function onDragEnd() {}

                const onDragDrop = () => {}

                const onDragLeave = function() {}

                return (
                    <Panel 
                        snippet="Button"
                        id="RootPanel"
                        class="root" 
                        hittest={false} 
                        ref={root} 
                        tooltip_text="#addon_game_name"
                        custom_tooltip="tootip_example"
                        custom_tooltip_params={{name: 'test'}}
                        className="root"
                        classList={{root: false}}
                        enabled={false}
                        visible={false}
                        checked={false}
                        hittestchildren={false}
                        acceptsfocus={false}
                        tabindex={0}
                        inputnamespace=""
                        draggable

                        onactivate={() => {}}
                        onDragStart={() => {}}
                        onDragEnd={onDragEnd}
                        onDragDrop={onDragDrop}
                        onDragEnter={function() {}}
                        onDragLeave={onDragLeave}
                        >
                    </Panel>
                );
            }
            
            render(() => <App />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });

    test('custom properties: dynamic', function () {
        const result = parseSolid(`
            import { render } from 'solid-panorama-runtime';

            function App() {
                let [enabled] = createSignal(false);
                let [myStyle] = createSignal('');

                return (
                    <Panel
                        class={myStyle()}
                        className={myStyle()}
                        hittest={enabled()}
                        enabled={enabled()}
                        visible={enabled()}
                        checked={enabled()}
                        >
                    </Panel>
                );
            }
            
            render(() => <App />, $('#app'));
        `);
        expect(result).toMatchSnapshot();
    });
});
