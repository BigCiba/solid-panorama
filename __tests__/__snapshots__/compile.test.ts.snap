// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile custom properties: dynamic 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function App() {
  let [enabled] = createSignal(false);
  let [myStyle] = createSignal('');
  return (() => {
    const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", {
      get ["class"]() {
        return myStyle();
      },
      get hittest() {
        return enabled();
      }
    }, null);
    (0, _solidPanoramaRuntime.effect)(_p$ => {
      const _v$ = myStyle(),
        _v$2 = myStyle(),
        _v$3 = enabled(),
        _v$4 = enabled(),
        _v$5 = enabled(),
        _v$6 = enabled();
      _v$ !== _p$._v$ && (_p$._v$ = (0, _solidPanoramaRuntime.setProp)(_el$, "class", _v$, _p$._v$));
      _v$2 !== _p$._v$2 && (_p$._v$2 = (0, _solidPanoramaRuntime.setProp)(_el$, "className", _v$2, _p$._v$2));
      _v$3 !== _p$._v$3 && (_p$._v$3 = (0, _solidPanoramaRuntime.setProp)(_el$, "hittest", _v$3, _p$._v$3));
      _v$4 !== _p$._v$4 && (_p$._v$4 = (0, _solidPanoramaRuntime.setProp)(_el$, "enabled", _v$4, _p$._v$4));
      _v$5 !== _p$._v$5 && (_p$._v$5 = (0, _solidPanoramaRuntime.setProp)(_el$, "visible", _v$5, _p$._v$5));
      _v$6 !== _p$._v$6 && (_p$._v$6 = (0, _solidPanoramaRuntime.setProp)(_el$, "checked", _v$6, _p$._v$6));
      return _p$;
    }, {
      _v$: undefined,
      _v$2: undefined,
      _v$3: undefined,
      _v$4: undefined,
      _v$5: undefined,
      _v$6: undefined
    });
    return _el$;
  })();
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(App, {}), $('#app'));"
`;

exports[`compile custom properties: initialize 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function App() {
  let root;
  function onDragEnd() {}
  const onDragDrop = () => {};
  const onDragLeave = function () {};
  return (() => {
    const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", {
      snippet: "Button",
      id: "RootPanel",
      "class": "root",
      hittest: false,
      hittestchildren: false,
      acceptsfocus: false,
      tabindex: 0,
      inputnamespace: "",
      draggable: true
    }, null);
    const _ref$ = root;
    typeof _ref$ === "function" ? (0, _solidPanoramaRuntime.use)(_ref$, _el$) : root = _el$;
    (0, _solidPanoramaRuntime.setProp)(_el$, "tooltip_text", "#addon_game_name");
    (0, _solidPanoramaRuntime.setProp)(_el$, "custom_tooltip", "tootip_example");
    (0, _solidPanoramaRuntime.setProp)(_el$, "custom_tooltip_params", {
      name: 'test'
    });
    (0, _solidPanoramaRuntime.setProp)(_el$, "className", "root");
    (0, _solidPanoramaRuntime.setProp)(_el$, "classList", {
      root: false
    });
    (0, _solidPanoramaRuntime.setProp)(_el$, "onactivate", () => {});
    (0, _solidPanoramaRuntime.setProp)(_el$, "onDragStart", () => {});
    (0, _solidPanoramaRuntime.setProp)(_el$, "onDragEnd", onDragEnd);
    (0, _solidPanoramaRuntime.setProp)(_el$, "onDragDrop", onDragDrop);
    (0, _solidPanoramaRuntime.setProp)(_el$, "onDragEnter", function () {});
    (0, _solidPanoramaRuntime.setProp)(_el$, "onDragLeave", onDragLeave);
    return _el$;
  })();
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(App, {}), $('#app'));"
`;

exports[`compile transform: attrs 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function HelloWorld() {
  const [ability, setAbility] = createSignal("");
  return (() => {
    const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", {}, null),
      _el$2 = (0, _solidPanoramaRuntime.createElement)("Panel", {}, _el$);
    (0, _solidPanoramaRuntime.setProp)(_el$2, "attrs", {
      name: 'robin'
    });
    return _el$;
  })();
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(HelloWorld, {}), $('#app'));"
`;

exports[`compile transform: dynamic 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
const RedThing = () => (() => {
  const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", {
    style: "color: red;"
  }, null);
  (0, _solidPanoramaRuntime.createTextNode)(\`Red Thing\`, _el$);
  return _el$;
})();
const GreenThing = () => (() => {
  const _el$3 = (0, _solidPanoramaRuntime.createElement)("Panel", {
    style: "color: green;"
  }, null);
  (0, _solidPanoramaRuntime.createTextNode)(\`Green Thing\`, _el$3);
  return _el$3;
})();
const BlueThing = () => (() => {
  const _el$5 = (0, _solidPanoramaRuntime.createElement)("Panel", {
    style: "color: blue;"
  }, null);
  (0, _solidPanoramaRuntime.createTextNode)(\`Blue Thing\`, _el$5);
  return _el$5;
})();
const options = {
  red: RedThing,
  green: GreenThing,
  blue: BlueThing
};
function HelloWorld() {
  const [selected, setSelected] = createSignal("red");
  return (0, _solidPanoramaRuntime.createComponent)(_solidPanoramaRuntime.Dynamic, {
    get component() {
      return options[selected()];
    }
  });
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(HelloWorld, {}), $('#app'));"
`;

exports[`compile transform: for each 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
var _web = require("solid-js/web");
var _solidJs = require("solid-js");
const App = () => {
  const [list, setList] = (0, _solidJs.createSignal)(['A']);
  function click() {
    setList([...list(), "B"]);
  }
  return (() => {
    const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", {}, null),
      _el$2 = (0, _solidPanoramaRuntime.createElement)("Button", {}, _el$);
    (0, _solidPanoramaRuntime.insert)(_el$, (0, _solidPanoramaRuntime.createComponent)(_solidPanoramaRuntime.For, {
      get each() {
        return list();
      },
      children: (item, index) => {
        return (() => {
          const _el$4 = (0, _solidPanoramaRuntime.createElement)("Label", {
            text: item
          }, null);
          (0, _solidPanoramaRuntime.setProp)(_el$4, "text", item);
          return _el$4;
        })();
      }
    }), _el$2);
    (0, _solidPanoramaRuntime.createTextNode)(\`Click\`, _el$2);
    (0, _solidPanoramaRuntime.setProp)(_el$2, "onClick", click);
    return _el$;
  })();
};
(0, _web.render)(() => (0, _solidPanoramaRuntime.createComponent)(App, {}), $('#app'));"
`;

exports[`compile transform: fragment 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function HelloWorld() {
  return [(0, _solidPanoramaRuntime.createElement)("Button", {}, null), (0, _solidPanoramaRuntime.createElement)("Button", {}, null), (0, _solidPanoramaRuntime.createElement)("Button", {}, null)];
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(HelloWorld, {}), $('#app'));"
`;

exports[`compile transform: normal 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function Item(props) {
  const [rootStyle, setRootStyle] = createSignal('root');
  let visible = true;
  let root, btnA;
  return (() => {
    const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", {
        id: "root"
      }, null),
      _el$2 = (0, _solidPanoramaRuntime.createElement)("Label", {
        text: "Testing"
      }, _el$),
      _el$3 = (0, _solidPanoramaRuntime.createElement)("Panel", {}, _el$),
      _el$4 = (0, _solidPanoramaRuntime.createElement)("Button", {}, _el$3),
      _el$5 = (0, _solidPanoramaRuntime.createElement)("Button", {}, _el$3),
      _el$6 = (0, _solidPanoramaRuntime.createElement)("Button", {}, _el$3);
    const _ref$ = root;
    typeof _ref$ === "function" ? (0, _solidPanoramaRuntime.use)(_ref$, _el$) : root = _el$;
    (0, _solidPanoramaRuntime.setProp)(_el$3, "className", "buttons");
    (0, _solidPanoramaRuntime.use)(a => btnA = a, _el$4);
    (0, _solidPanoramaRuntime.setProp)(_el$5, "visible", visible);
    (0, _solidPanoramaRuntime.effect)(_$p => (0, _solidPanoramaRuntime.setProp)(_el$, "className", rootStyle().join(' '), _$p));
    return _el$;
  })();
}
function HelloWorld() {
  return (() => {
    const _el$7 = (0, _solidPanoramaRuntime.createElement)("Panel", {}, null);
    (0, _solidPanoramaRuntime.insert)(_el$7, (0, _solidPanoramaRuntime.createComponent)(Item, {}), null);
    (0, _solidPanoramaRuntime.insert)(_el$7, (0, _solidPanoramaRuntime.createComponent)(Item, {
      show: true
    }), null);
    return _el$7;
  })();
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(HelloWorld, {}), $('#app'));"
`;

exports[`compile transform: snippet 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function HelloWorld() {
  return (0, _solidPanoramaRuntime.createElement)("Button", {
    snippet: "templateA"
  }, null);
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(HelloWorld, {}), $('#app'));"
`;

exports[`compile transform: spread operator 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function Item(props) {
  return (() => {
    const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", props, null);
    (0, _solidPanoramaRuntime.spread)(_el$, props, false);
    return _el$;
  })();
}
function HelloWorld() {
  return (0, _solidPanoramaRuntime.createComponent)(Item, {});
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(HelloWorld, {}), $('#app'));"
`;

exports[`compile transform: style 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function Item(props) {
  return (() => {
    const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", {
      get ["class"]() {
        return props.style;
      }
    }, null);
    (0, _solidPanoramaRuntime.effect)(_$p => (0, _solidPanoramaRuntime.setProp)(_el$, "class", props.style, _$p));
    return _el$;
  })();
}
function HelloWorld() {
  return (() => {
    const _el$2 = (0, _solidPanoramaRuntime.createElement)("Panel", {}, null),
      _el$3 = (0, _solidPanoramaRuntime.createElement)("Button", {}, _el$2),
      _el$4 = (0, _solidPanoramaRuntime.createElement)("Button", {
        style: "width:12px; height: 12px;"
      }, _el$2);
    (0, _solidPanoramaRuntime.setProp)(_el$3, "style", {
      width: '12px',
      height: 12
    });
    (0, _solidPanoramaRuntime.insert)(_el$2, (0, _solidPanoramaRuntime.createComponent)(Item, {
      style: "red"
    }), null);
    return _el$2;
  })();
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(HelloWorld, {}), $('#app'));"
`;

exports[`compile transform: textNode 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function HelloWorld() {
  return (() => {
    const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", {}, null),
      _el$2 = (0, _solidPanoramaRuntime.createElement)("Panel", {}, _el$),
      _el$4 = (0, _solidPanoramaRuntime.createElement)("Panel", {}, _el$),
      _el$6 = (0, _solidPanoramaRuntime.createElement)("Panel", {}, _el$),
      _el$7 = (0, _solidPanoramaRuntime.createElement)("Label", {
        text: "Welcome"
      }, _el$6),
      _el$8 = (0, _solidPanoramaRuntime.createTextNode)(\`#addon_game_name\`, _el$6),
      _el$9 = (0, _solidPanoramaRuntime.createElement)("Label", {
        text: "(～￣▽￣)～"
      }, _el$6);
    (0, _solidPanoramaRuntime.createTextNode)(\`Welcome My Game\`, _el$2);
    (0, _solidPanoramaRuntime.createTextNode)(\`<strong>Welcome</strong> My Game\`, _el$4);
    return _el$;
  })();
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(HelloWorld, {}), $('#app'));"
`;

exports[`compile transform: vars 1`] = `
""use strict";

var _solidPanoramaRuntime = require("solid-panorama-runtime");
function HelloWorld() {
  const [ability, setAbility] = createSignal("");
  return (() => {
    const _el$ = (0, _solidPanoramaRuntime.createElement)("Panel", {}, null),
      _el$2 = (0, _solidPanoramaRuntime.createElement)("Panel", {
        vars: {
          name: 'robin'
        }
      }, _el$),
      _el$3 = (0, _solidPanoramaRuntime.createElement)("Panel", {
        get vars() {
          return {
            name: 'robin',
            ability: ability()
          };
        },
        dialogVariables: {}
      }, _el$);
    (0, _solidPanoramaRuntime.setProp)(_el$2, "vars", {
      name: 'robin'
    });
    (0, _solidPanoramaRuntime.setProp)(_el$3, "dialogVariables", {});
    (0, _solidPanoramaRuntime.effect)(_$p => (0, _solidPanoramaRuntime.setProp)(_el$3, "vars", {
      name: 'robin',
      ability: ability()
    }, _$p));
    return _el$;
  })();
}
(0, _solidPanoramaRuntime.render)(() => (0, _solidPanoramaRuntime.createComponent)(HelloWorld, {}), $('#app'));"
`;
